File: app/src/main/java/pl/pola_app/ui/adapter/ProductList.java
package pl.pola_app.ui.adapter;

import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import com.crashlytics.android.answers.Answers;
import com.crashlytics.android.answers.CustomEvent;

import org.parceler.Parcels;

import java.util.ArrayList;
import java.util.List;

import pl.pola_app.BuildConfig;
import pl.pola_app.model.SearchResult;

public class ProductList {
    @NonNull private final List<SearchResult> searchResults;
    @Nullable private OnProductListChanged onProductListChanged;

    public static ProductList create(@Nullable Bundle bundle) {
        final List<SearchResult> searchResults;
        if (bundle != null) {
            searchResults = Parcels.unwrap(bundle.getParcelable(SearchResult.class.getName()));
        } else {
            searchResults = new ArrayList<>();
        }

        return new ProductList(searchResults);
    }

    private ProductList(@NonNull final List<SearchResult> searchResults) {
        this.searchResults = searchResults;
    }

    public void setOnProductListChanged(@Nullable OnProductListChanged onProductListChanged) {
        this.onProductListChanged = onProductListChanged;
    }


    public void writeToBundle(@NonNull final Bundle bundle) {
        bundle.putParcelable(SearchResult.class.getName(), Parcels.wrap(searchResults));
    }

    public void addProduct(SearchResult searchResult) {
        if (searchResults.size() > 0 && searchResults.get(0) == null) {
            if (BuildConfig.USE_CRASHLYTICS) {
                Answers.getInstance().logCustom(new CustomEvent("addProduct")
                        .putCustomAttribute("good", "true"));
            }
            searchResults.set(0, searchResult);
        } else {
            if (BuildConfig.USE_CRASHLYTICS) {
                Answers.getInstance().logCustom(new CustomEvent("addProduct")
                        .putCustomAttribute("good", "false"));
            }
            searchResults.add(0, searchResult);
        }
        notifyOnChanged();
    }

    public void createProductPlaceholder() {
        searchResults.add(0, null);
        notifyOnChanged();
    }

    public void removeProductPlaceholder() {
        if (searchResults.size() > 0 && searchResults.get(0) == null) {
            searchResults.remove(0);
            notifyOnChanged();
        }
    }

    public boolean itemExists(String code) {
        for (SearchResult p : searchResults) {
            if (p != null) {
                if (p.code.equals(code)) {
                    searchResults.remove(p);
                    searchResults.add(0, p);
                    notifyOnChanged();
                    return true;
                }
            }
        }

        return false;
    }

    public int size() {
        return searchResults.size();
    }

    public SearchResult get(int position) {
        return searchResults.get(position);
    }

    private void notifyOnChanged() {
        if (onProductListChanged != null) {
            onProductListChanged.onChanged();
        }
    }

}

File: app/src/main/java/pl/pola_app/ui/adapter/ProductList.kt
package pl.pola_app.ui.adapter

import android.os.Bundle
import org.parceler.Parcels
import pl.pola_app.model.SearchResult
import java.util.*

class ProductList private constructor(private val searchResults: MutableList<SearchResult?>) {
    private var onProductListChanged: OnProductListChanged? = null
    fun setOnProductListChanged(onProductListChanged: () -> Unit) {
        this.onProductListChanged = object : OnProductListChanged{
            override fun onChanged() {
                onProductListChanged.invoke()
            }

        }
    }

    fun writeToBundle(bundle: Bundle) {
        bundle.putParcelable(
            SearchResult::class.java.name, Parcels.wrap<List<SearchResult?>>(
                searchResults
            )
        )
    }

    fun addProduct(searchResult: SearchResult?) {
        if (searchResults.size > 0 && searchResults[0] == null) {
            searchResults[0] = searchResult
        } else {
            searchResults.add(0, searchResult)
        }
        notifyOnChanged()
    }

    fun createProductPlaceholder() {
        searchResults.add(0, null)
        notifyOnChanged()
    }

    fun removeProductPlaceholder() {
        if (searchResults.size > 0 && searchResults[0] == null) {
            searchResults.removeAt(0)
            notifyOnChanged()
        }
    }

    fun itemExists(code: String): Boolean {
        for (p in searchResults) {
            if (p != null) {
                if (p.code == code) {
                    searchResults.remove(p)
                    searchResults.add(0, p)
                    notifyOnChanged()
                    return true
                }
            }
        }
        return false
    }

    fun size(): Int {
        return searchResults.size
    }

    operator fun get(position: Int): SearchResult? {
        return searchResults[position]
    }

    private fun notifyOnChanged() {
        onProductListChanged?.onChanged()
    }

    companion object {
        @kotlin.jvm.JvmStatic
        fun create(bundle: Bundle?): ProductList {
            val searchResults: MutableList<SearchResult?> = if (bundle != null) {
                Parcels.unwrap(
                    bundle.getParcelable(
                        SearchResult::class.java.name
                    )
                )
            } else {
                ArrayList()
            }
            return ProductList(searchResults)
        }
    }
}